dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
+{
+    return m_outerList.GetSize();
+}
+
+CStringW CAutoComplete::GetItemText(INT iItem)
+{
+    if (iItem == -1 || iItem >= m_outerList.GetSize())
+        return L"";
+    return m_outerList[iItem];
+}
+
+CStringW CAutoComplete::GetEditText()
+{
+    BSTR bstrText = NULL;
+    CStringW strText;
+    if (m_hwndEdit.GetWindowTextW(bstrText))
+    {
+        strText = bstrText;
+        ::SysFreeString(bstrText);
+    }
+    return strText;
+}
+
+VOID CAutoComplete::SetEditText(LPCWSTR pszText)
+{
+    m_bInSetText = TRUE; // don't hide drop-down
+    m_hwndEdit.SetWindowTextW(pszText);
+    m_bInSetText = FALSE;
+}
+
+CStringW CAutoComplete::GetStemText()
+{
+    CStringW strText = GetEditText();
+    INT ich = strText.ReverseFind(L'\\');
+    if (ich == -1)
+        return L""; // no stem
+    return strText.Left(ich + 1);
+}
+
+VOID CAutoComplete::SetEditSel(INT ich0, INT ich1)
+{
+    m_hwndEdit.SendMessageW(EM_SETSEL, ich0, ich1);
+}
+
+VOID CAutoComplete::ShowDropDown()
+{
+    if (!m_hWnd || !CanAutoSuggest())
+        return;
+
+    INT cItems = GetItemCount();
+    if (cItems == 0 || ::GetFocus() != m_hwndEdit || IsComboBoxDropped())
+    {
+        // hide the drop-down if necessary
+        HideDropDown();
+        return;
+    }
+
+    RepositionDropDown();
+}
+
+VOID CAutoComplete::HideDropDown()
+{
+    ShowWindow(SW_HIDE);
+}
+
+VOID CAutoComplete::SelectItem(INT iItem)
+{
+    m_hwndList.SetCurSel(iItem);
+    if (iItem != -1)
+        m_hwndList.EnsureVisible(iItem, FALSE);
+}
+
+VOID CAutoComplete::DoAutoAppend()
+{
+    if (!CanAutoAppend()) // can we auto-append?
+        return;
+
+    CStringW strText = GetEditText(); // get the text
+    if (strText.IsEmpty())
+        return;
+
+    INT cItems = m_innerList.GetSize(); // get the number of items
+    if (cItems == 0)
+        return; // don't append
+
+    // get the common string
+    CStringW strCommon;
+    for (INT iItem = 0; iItem < cItems; ++iItem)
+    {
+        const CString& strItem = m_innerList[iItem]; // get the text of the item
+
+        if (iItem == 0) // the first item
+        {
+            strCommon = strItem; // store the text
+            continue;
+        }
+
+        for (INT ich = 0; ich < strCommon.GetLength(); ++ich)
+        {
+            if (ich < strItem.GetLength() && strCommon[ich] != strItem[ich])
+            {
+                strCommon = strCommon.Left(ich); // shrink the common string
+                break;
+            }
+        }
+    }
+
+    if (strCommon.IsEmpty() || strCommon.GetLength() <= strText.GetLength())
+        return; // no suggestion
+
+    // append suggestion
+    INT cchAppend = strCommon.GetLength() - strText.GetLength();
+    CStringW strAppend = strCommon.Right(cchAppend);
+    strText += strAppend;
+    SetEditText(strText);
+
+    // select the appended
+    INT ich0 = strText.GetLength();
+    INT ich1 = ich0 + cchAppend;
+    SetEditSel(ich0, ich1);
+}
+
+// go back a word ([Ctrl]+[Backspace])
```

Comment:
I always wondered: Is this functionality an actual default one present in all edit text controls? (it seems to appear so in Windows, at least 10)

Reply:
No. Notepad of Win2k3 generates a tofu block on Ctrl+BackSpace.

![VirtualBox_Windows Server 2003_09_03_2021_12_30_35](https://user-images.githubusercontent.com/2107452/110414544-5f48e680-80d3-11eb-96fd-d41a1eacc0f8.png)

Reply:
I can observe the window procedure change of the EDIT control in auto-completion textboxes.